File: problems/1/198 House Robber/rob.py
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        n = len(nums)
        if n == 1:
            return nums[0]
        sol = [nums[0], max(nums[:2])]
        for i in range(2, n):
            sol.append(max(nums[i] + sol[i-2], sol[i-1]))
        return sol[-1]

sol = Solution()
print(sol.rob([1,2,3,1]))

File: problems/1/198 House Robber/rob.rs
struct Solution;

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        if n == 0 {
            return 0;
        } else if n == 1 {
            return nums[0];
        }
        let mut sol = vec![nums[0]; n];
        sol[0] = nums[0];
        sol[1] = nums[0].max(nums[1]);
        for i in 2..n {
            sol[i] = sol[i-1].max(sol[i-2] + nums[i]);
        }
        sol[n-1]
    }
}

fn main() {
    println!("{}", Solution::rob(vec![]));
}